drivers/gpio/gpio_renesas_ra2.c

```diff
@@ -0,0 +1,378 @@
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_utils.h>
+
+#include <zephyr/drivers/pinctrl/pinctrl_ra2.h>
+#include <zephyr/drivers/interrupt_controller/intc_ra2_icu.h>
+#include <zephyr/drivers/gpio/gpio_renesas_ra2.h>
+
+#define DT_DRV_COMPAT renesas_ra2_gpio
+
+#define PCNTR1_OFF		0x0
+#define		PODR_OFF	0x0
+#define		PDR_OFF		0x2
+#define PCNTR2_OFF		0x4
+#define		EIDR_OFF	0x4
+#define		PIDR_OFF	0x6
+#define PCNTR3_OFF		0x8
+#define		PORR_OFF	0x8
+#define		POSR_OFF	0xa
+#define PCNTR4_OFF		0xc
+#define		EORR_OFF	0xc
+#define		EOSR_OFF	0xe
+
+typedef struct gpio_irq_item {
+	gpio_pin_t	pin;
+	uint8_t		irq;
+} gpio_irq_item_t;
+
+struct gpio_ra_config {
+	/* gpio_driver_config needs to be first */
+	struct gpio_driver_config common;
+	mm_reg_t  base;
+	const gpio_irq_item_t *irqs;
+	uint8_t ioport_id;
+	uint8_t num_irqs;
+};
+
+struct gpio_ra_data {
+	/* gpio_driver_data needs to be first */
+	struct gpio_driver_data common;
+	/* port ISR callback routine address */
+	sys_slist_t callbacks;
+};
+
+struct port_irq_event {
+	uint8_t event1;
+	uint8_t event2;
+};
+
+struct port_irq_data {
+	struct icu_event	*event;
+	const struct device	*gpio_port;
+	gpio_pin_t		pin;
+};
+
+static K_MUTEX_DEFINE(lock);
+
+const struct port_irq_event port_irq_events[] = {
+	{ ICU_EVENT_GROUP0_PORT_IRQ0, ICU_EVENT_GROUP4_PORT_IRQ0, },
+	{ ICU_EVENT_GROUP1_PORT_IRQ1, ICU_EVENT_GROUP5_PORT_IRQ1, },
+	{ ICU_EVENT_GROUP2_PORT_IRQ2, ICU_EVENT_GROUP6_PORT_IRQ2, },
+	{ ICU_EVENT_GROUP3_PORT_IRQ3, ICU_EVENT_GROUP7_PORT_IRQ3, },
+	{ ICU_EVENT_GROUP4_PORT_IRQ4, 0, },
+	{ ICU_EVENT_GROUP5_PORT_IRQ5, 0, },
+	{ ICU_EVENT_GROUP6_PORT_IRQ6, 0, },
+	{ ICU_EVENT_GROUP7_PORT_IRQ7, 0, },
+};
+
+struct port_irq_data irq_datas[ARRAY_SIZE(port_irq_events)];
+
+static int gpio_ra_configure(const struct device *dev, gpio_pin_t pin,
+			      gpio_flags_t flags)
+{
+	const struct gpio_ra_config *cfg = dev->config;
+
+	if ((flags & (GPIO_INPUT | GPIO_OUTPUT)) ==
+			(GPIO_INPUT | GPIO_OUTPUT)) {
+		return -ENOTSUP;
+	}
+
+	if ((flags & GPIO_PULL_DOWN) != 0U) {
+		return -ENOTSUP;
+	}
+
+	if ((BIT(pin) & cfg->common.port_pin_mask) != 0) {
+		pinctrl_soc_pin_t pfs_cache = {
+			RA_PIN_FLAGS_PIN(pin) |
+			RA_PIN_FLAGS_PORT(cfg->ioport_id),
+		};
+
+		if (flags & GPIO_OUTPUT) {
+			pfs_cache |= RA_PIN_FLAGS_PDR;
+		}
+
+		if (flags & GPIO_OUTPUT_INIT_HIGH) {
+			pfs_cache |= RA_PIN_FLAGS_PODR;
+		}
+
+		if (flags & GPIO_PULL_UP) {
+			pfs_cache |= RA_PIN_FLAGS_PCR;
+		}
+
+		if (flags & GPIO_OPEN_DRAIN) {
+			pfs_cache |= RA_PIN_FLAGS_NCODR;
+		}
+
+		return pinctrl_configure_pins(&pfs_cache, 1, PINCTRL_REG_NONE);
+	}
+
+	return -ENODEV;
+}
+
+static int gpio_ra_port_get_raw(const struct device *port, uint32_t *value)
+{
+	const struct gpio_ra_config *cfg = port->config;
+
+	*value = sys_read16(cfg->base + PIDR_OFF) & cfg->common.port_pin_mask;
+
+	return 0;
+}
+
+static int gpio_ra_port_set_masked_raw(const struct device *port,
```

Comment:
`gpio_port_set_masked_raw()` may also set a pin to low physical level. Does the current implementation support this?